openedx_events/event_bus/avro/deserializer.py

```diff
@@ -0,0 +1,89 @@
+"""
+Deserialize Avro records to events that can be sent with OpenEdxPublicSignals.
+"""
+import json
+from datetime import datetime
+
+import attr
+from opaque_keys.edx.keys import CourseKey
+
+from .custom_serializers import CourseKeyAvroSerializer, DatetimeAvroSerializer
+from .schema import schema_from_signal
+from .types import PYTHON_TYPE_TO_AVRO_MAPPING
+
+DEFAULT_DESERIALIZERS = {
+    datetime: DatetimeAvroSerializer.deserialize,
+    CourseKey: CourseKeyAvroSerializer.deserialize,
+}
+
+
+def _deserialized_avro_dict_to_object(data: dict, data_type, deserializers=None):
+    """
+    Convert dictionary entry into an instance of data_type.
+
+    Used to convert messages from an AvroDeserializer into events that can be sent by the
+    appropriate signal instance
+
+    Arguments:
+        data: Dictionary returned from AvroDeserializer
+        data_type: Desired Python data type, eg `str`, `CourseKey`, `CourseEnrollmentData`
+
+    Returns:
+        An instance of data_type
+    """
+    param_deserializers = deserializers or {}
+    all_deserializers = {**DEFAULT_DESERIALIZERS, **param_deserializers}
+
+    if deserializer := all_deserializers.get(data_type, None):
+        return deserializer(data)
+    elif data_type in PYTHON_TYPE_TO_AVRO_MAPPING:
+        return data
+    elif hasattr(data_type, "__attrs_attrs__"):
+        transformed = {}
+        for attribute in data_type.__attrs_attrs__:
+            if attribute.name in data:
+                sub_data = data[attribute.name]
+                if sub_data or attribute.default is attr.NOTHING:
+                    transformed[attribute.name] = _deserialized_avro_dict_to_object(sub_data,
+                                                                                    attribute.type,
+                                                                                    deserializers=deserializers)
+
+        return data_type(**transformed)
+    raise TypeError(
+        f"Unable to deserialize {data_type} data, please add extension for custom data type"
+    )
+
+
+def _record_dict_to_event_data(signal, avro_record, deserializers=None):
+    return {data_key: _deserialized_avro_dict_to_object(avro_record[data_key], data_type, deserializers)
+            for data_key, data_type in signal.init_data.items()}
+
+
+class AvroSignalDeserializer:
+    """
+    Class to deserialize Avro records into events that can be sent by self.signal.
+
+    The schema_string and from_dict methods can be used by the confluent_kafka.AvroDeserializer class when creating
+    an event consumer.
```

Comment:
What about something like the following:
```suggestion
    The ``schema_string`` and ``from_dict`` methods are the API for this deserializer, and could potentially be used by multiple event bus technologies, but the names come directly from the confluent_kafka.AvroDeserializer class that is used when creating
    an event consumer.
```